File: Full-System Simulation/gem5/src/dev/PIMAdder.py


from m5.params import *
from m5.proxy import *
from m5.SimObject import SimObject
from m5.objects.PIMKernel import PIMKernel

class PIMAdder(PIMKernel):
    type = 'PIMAdder'
    cxx_header = "dev/pimadder.hh"
    name = Param.String("ADDer","PIM Unit name.")
    latency = Param.Int("1", "PIM Unit computation delay cycles.")
    input=Param.Int(4, "num of inputs")
    output=Param.Int(1, "num of outputs")
    

File: Full-System Simulation/gem5/src/dev/pimadder.cc
#include "dev/pimkernel.hh"
#include "dev/pimadder.hh"
#include "params/PIMAdder.hh"
#include <assert.h>


PIMAdder::PIMAdder(const Params *p):
    PIMKernel(p)
{
    
}


PIMAdder::~PIMAdder(){
    
}

PIMAdder *
PIMAdderParams::create()
{
    return new PIMAdder(this);
}


void
PIMAdder::ADD(){
    uint64_t result=0;
    assert(inputReady());
    
    for(int i=0;i<_input;i++){

        result+=regs[i];
    }
    
    regs[_input]=result;
    
    data_ready[_input]=true;
}
